libcudacxx/include/cuda/std/__numeric/saturation_arithmetic.h

```diff
@@ -526,13 +484,74 @@ class __mul_sat
     return __result;
   }
 
+public:
   template <class _Tp>
-  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static _Tp __impl_generic(_Tp __x, _Tp __y) noexcept
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __impl_generic(_Tp __x, _Tp __y) noexcept
   {
-    return __impl_constexpr(__x, __y);
+    if (__x == 0 || __y == 0)
+    {
+      return _Tp{};
+    }
+
+    if constexpr (_CCCL_TRAIT(is_signed, _Tp))
+    {
+      if (__x == -1)
+      {
+        if (__y == _CUDA_VSTD::numeric_limits<_Tp>::min())
+        {
+          return _CUDA_VSTD::numeric_limits<_Tp>::max();
+        }
+        return -__y;
```

Comment:
unary minus is subject to integer promotion, so this should `static_cast` to `_Tp`